networkit/cpp/graph/GraphBuilder.cpp

```diff
@@ -12,86 +12,163 @@
+        if (weighted)
+            inEdgeWeightsPerThread[i].resize(max_threads);
+    }
 }
 
-index GraphBuilder::indexInOutEdgeArray(node u, node v) const {
-    for (index i = 0; i < outEdges[u].size(); i++) {
-        node x = outEdges[u][i];
-        if (x == v) {
+index GraphBuilder::indexInOutEdgeArrayPerThread(node u, node v) const {
+    int max_threads = omp_get_max_threads();
+    int thread_num = omp_get_thread_num();
+    auto &edges = outEdgesPerThread[thread_num][u % max_threads];
+    for (index i = 0; i < edges.size(); ++i) {
+        if (edges[i].source == u && edges[i].destination == v)
             return i;
-        }
     }
     return none;
 }
 
-index GraphBuilder::indexInInEdgeArray(node u, node v) const {
-    assert(isDirected());
-    for (index i = 0; i < inEdges[u].size(); i++) {
-        node x = inEdges[u][i];
-        if (x == v) {
+index GraphBuilder::indexInInEdgeArrayPerThread(node u, node v) const {
+    int max_threads = omp_get_max_threads();
+    int thread_num = omp_get_thread_num();
+    auto &edges = inEdgesPerThread[thread_num][u % max_threads];
+    for (index i = 0; i < edges.size(); ++i) {
+        if (edges[i].source == u && edges[i].destination == v)
             return i;
-        }
     }
     return none;
 }
 
 node GraphBuilder::addNode() {
-    outEdges.emplace_back();
-    if (weighted) {
-        outEdgeWeights.emplace_back();
-    }
-    if (directed) {
-        inEdges.emplace_back();
-        if (weighted) {
-            inEdgeWeights.emplace_back();
-        }
-    }
     return n++;
 }
 
-void GraphBuilder::addHalfOutEdge(node u, node v, edgeweight ew) {
-    assert(indexInOutEdgeArray(u, v) == none);
-    outEdges[u].push_back(v);
-    if (weighted) {
-        outEdgeWeights[u].push_back(ew);
+void GraphBuilder::addHalfEdge(index a, index b, edgeweight ew) {
+    if (!useWholeEdges) {
+        addHalfOutEdge(a, b, ew);
+    } else {
+        if (directed) {
+            addHalfOutEdge(a, b, ew);
+            addHalfInEdge(b, a);
+        } else {
+            addHalfOutEdge(a, b, ew);
+            if (a != b)
+                addHalfOutEdge(b, a, ew);
+        }
     }
-    if (u == v) {
+    if (a == b) {
 #pragma omp atomic
         selfloops++;
     }
 }
 
-void GraphBuilder::addHalfInEdge(node u, node v, edgeweight ew) {
-    assert(indexInInEdgeArray(u, v) == none);
-    inEdges[u].push_back(v);
+void GraphBuilder::addHalfOutEdge(node u, node v, edgeweight ew) {
+    outEdgesPerThread[omp_get_thread_num()][u % omp_get_max_threads()].emplace_back(HalfEdge{u, v});
     if (weighted) {
-        inEdgeWeights[u].push_back(ew);
+        outEdgeWeightsPerThread[omp_get_thread_num()][u % omp_get_max_threads()].emplace_back(ew);
     }
-    if (u == v) {
-#pragma omp atomic
-        selfloops++;
+}
+
+void GraphBuilder::addHalfInEdge(node u, node v, edgeweight ew) {
+    inEdgesPerThread[omp_get_thread_num()][u % omp_get_max_threads()].emplace_back(HalfEdge{u, v});
```

Comment:
Add a proper constructor to `HalfEdge` if needed and use
```suggestion
    inEdgesPerThread[omp_get_thread_num()][u % omp_get_max_threads()].emplace_back(u, v);
```